### PR TITLE
[KernelGen][Nvidia] Add soft_margin_loss_backward operator with Triton kernel

### PR DESCRIPTION
**Pull request by XDYuanzhuLee**
## Summary
Adds a Triton kernel for `soft_margin_loss_backward`. Backward pass for soft margin loss, a two-class classification logistic loss between input and target tensors.

## Testing
- Validated against reference on device via `to_reference(inp, True)`
- Tested on: Nvidia, Tianshu, Muxi, Ascend, Hygon

## Performance
Test command: `pytest benchmark/test_soft_margin_loss_backward.py --level core` (NVIDIA H20)

### soft_margin_loss_backward

| dtype | Size | Torch Latency (ms) | Gems Latency (ms) | Speedup |
|-------|------|--------------------|-------------------|---------|
| float16 | [128, 128] | 0.021856 | 0.045408 | 0.481 |
| float16 | [512, 512] | 0.040544 | 0.051728 | 0.784 |
| float16 | [1024, 1024] | 0.198016 | 0.085040 | 2.329 |
| float16 | [2048, 2048] | 2.523648 | 0.398784 | 6.327 |
| float16 | [4096, 4096] | 22.333440 | 3.205632 | 6.967 |
| float32 | [128, 128] | 0.024960 | 0.047616 | 0.524 |
| float32 | [512, 512] | 0.059232 | 0.048288 | 1.227 |
| float32 | [1024, 1024] | 0.279680 | 0.102336 | 2.733 |
| float32 | [2048, 2048] | 3.508224 | 0.518656 | 6.764 |
| float32 | [4096, 4096] | 31.459328 | 4.179968 | 7.526 |
| bfloat16 | [128, 128] | 0.038016 | 0.047616 | 0.798 |
| bfloat16 | [512, 512] | 0.055296 | 0.053936 | 1.025 |
| bfloat16 | [1024, 1024] | 0.215040 | 0.088576 | 2.427 |
| bfloat16 | [2048, 2048] | 2.645504 | 0.411136 | 6.434 |
| bfloat16 | [4096, 4096] | 23.349248 | 3.271680 | 7.137 |

| Operator | Arithmetic Mean Speedup |
|----------|------------------------|
| soft_margin_loss_backward | **4.874** |

## Multi-backend Testing
| Backend | Accuracy Test | Speedup (mean) | Notes |
|---|---|---|---|
| Nvidia (H20) | PASS (15 cases) | 4.874 | Primary |
| Tianshu | PASS | 4.782 | — |
| Muxi | PASS | 0.744 | — |
| Ascend | PASS | — | benchmark not run |
| Hygon | PASS | 4.373 | — |

## Files Changed
- `src/flag_gems/ops/soft_margin_loss_backward.py`: Triton kernel implementation
- `tests/test_soft_margin_loss_backward.py`: Accuracy test
- `benchmark/test_soft_margin_loss_backward.py`: Performance benchmark
- `src/flag_gems/ops/__init__.py`: Register import and `__all__`
- `src/flag_gems/__init__.py`: Register to `_FULL_CONFIG`
- `conf/operators.yaml`: Add operator entry (kind: Math, stage: alpha 5.1)

### CODE REVIEW

File: src/flag_gems/ops/soft_margin_loss_backward.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+        if reduction in (0, 1, 2):
+            return reduction
+        raise ValueError(f"Invalid reduction int: {reduction}")
+    raise ValueError(f"Unsupported reduction type: {type(reduction)}")
+
+
+def _check_tensors(
+    grad_output: torch.Tensor, self_tensor: torch.Tensor, target: torch.Tensor
+):
+    if not (grad_output.is_cuda and self_tensor.is_cuda and target.is_cuda):
```
Comment (Dongxu-H) [inline src/flag_gems/ops/soft_margin_loss_backward.py]: is_cuda is invalid for non NV chips

Comment (XDYuanzhuLee) [reply src/flag_gems/ops/soft_margin_loss_backward.py]: Fixed: removed the .is_cuda check from _check_tensors. The same-device validation and kernel launch provide sufficient guards across all backends (Nvidia, Metax, Ascend, Hygon, etc.).

**Comment by w1120029931-bit** [comment]
please resolve conflicts

**Comment by XDYuanzhuLee** [comment]
Fixed in the latest push. The conflict is resolved and the PR is mergeable now. Validation: current CI checks are passing.

**Review by Dongxu-H — COMMENTED** [review]
## FlagGems Operator 贡献审查报告

**Operator**: `soft_margin_loss_backward`
**PR**: #3726

### 📊 变更统计
- 新增行数: 195
- 删除行数: 0  
- 变更文件: 6

### ✅ 检查清单

| 检查项 | 状态 |
|--------|------|
| 实现文件 (`src/flag_gems/ops/soft_margin_loss_backward.py`) | ✅ 存在 |
| 测试文件 (`tests/test_soft_margin_loss_backward.py`) | ✅ 存在 |
| Benchmark 文件 | ✅ 存在 |
| 配置注册 (`operators.yaml`) | ✅ 存在 |
| 模块注册 (`__init__.py`) | ✅ 存在 |
| KernelGen 标注 | ✅ 存在 |
| Triton JIT 装饰器 | ✅ 存在 |
| Metax Backend | ⚠️ 未实现 |

### 📝 审查意见

**代码结构**: 实现文件结构正确

**测试覆盖**: 有测试文件

**性能测试**: 有 benchmark

---

*此审查由自动化脚本根据 FlagGems Operator 贡献规范生成。*
*如有问题请联系维护者进行详细审查。*

**Review by Dongxu-H — COMMENTED** [review]
## FlagGems Operator Contribution Review

**Operator**: `soft_margin_loss_backward`  
**PR**: #3726

### 📊 Change Statistics
| Metric | Value |
|--------|-------|
| Additions | 195 |
| Deletions | 0 |
| Files Changed | 6 |

### 📋 22-Point Checklist

| # | Requirement | Status |
|---|-------------|--------|
| 1 | Implementation file path correctness | ✅ Correct: `src/flag_gems/ops/soft_margin_loss_backward.py` |
| 2 | KernelGen attribution in header | ✅ Present |
| 3 | Import order (stdlib → third-party → internal) | ⚠️ Could not verify |
| 4 | Function signature alignment with PyTorch API | ⚠️ Needs manual verification |
| 5 | Type support documentation | ✅ Type handling present |
| 6 | Triton kernel decorators (`@triton.jit`, etc.) | ⚠️ `@triton.jit` present, verify other decorators |
| 7 | Test file existence | ✅ Present: `tests/test_soft_margin_loss_backward.py` |
| 8 | Test coverage (shapes, dtypes) | ✅ Shape and dtype tests present |
| 9 | Benchmark file existence | ✅ Present: `benchmark/test_soft_margin_loss_backward.py` |
| 10 | Benchmark implementation correctness | ✅ Benchmark implemented |
| 11 | Config registration (`operators.yaml`) | ✅ Registered in `operators.yaml` |
| 12 | Module registration (`__init__.py`) | ✅ Registered in `__init__.py` |
| 13 | Ops module registration (`ops/__init__.py`) | ✅ Registered in `ops/__init__.py` |
| 14 | Logger namespace (`flag_gems.*`) | ⚠️ Missing `flag_gems.` prefix |
| 15 | Error handling | ✅ Error handling present |
| 16 | Numerical stability patterns | ⚠️ Needs domain expert review |
| 17 | Memory efficiency | ⚠️ Needs profiling |
| 18 | Edge case tests | ⚠️ Needs verification |
| 19 | Backend support (Metax) | ❌ Not implemented |
| 20 | Documentation/Docstrings | ⚠️ Needs manual review |
| 21 | Code style consistency | ⚠️ Run linter manually |
| 22 | Integration verification | ⚠️ Verify end-to-end |

### 📈 Summary

| Category | Count |
|----------|-------|
| ✅ Pass | 11 |
| ⚠️ Warning | 10 |
| ❌ Fail | 1 |

**Overall**: ⚠️ Needs Attention

---

*This review was generated by an automated script following the FlagGems Operator Contribution Guidelines.*
*For detailed code review, please consult a domain expert.*

**Review by Dongxu-H — COMMENTED** [review]
## 💡 审查建议

建议:
- ⚠️ 建议在flag_gems/__init__.py注册

**Review by Dongxu-H — COMMENTED** [review]
## 💡 审查建议

建议:
- ⚠️ 建议在flag_gems/__init__.py注册

**Review by Dongxu-H — COMMENTED** [review]
## 💡 审查建议

建议:
- ⚠️ 建议在flag_gems/__init__.py注册

**Review by Caeruleann — DISMISSED** [review]

**Comment by Dongxu-H** [comment]
## 代码审查结果 - PR #3726

### ✅ 通过的检查

1. **CI 工作流状态**：`python-op` 和 `code-style` 通过
2. **合并冲突**：无冲突 (MERGEABLE)
3. **配置一致性**：`conf/operators.yaml` 排序正确（`soft_margin_loss_backward` 在 `soft_margin_loss` 之后，`softmax` 之前）
4. **Kernel 注册**：`__init__.py` 导入和注册位置正确（按字母顺序）
5. **Logger 命名**：正确使用 `logging.getLogger(__name__)`（主文件夹位置）
6. **测试覆盖**：
   - 多种 shape 测试 ✓
   - 多种 dtype 测试 ✓
   - 三种 reduction 模式测试 (none/mean/sum) ✓

7. **实现逻辑**：
   - 正确的 backward 公式：`grad_output * (-target) / (1 + exp(target * self))` ✓
   - 正确处理非连续 tensor ✓
   - 正确处理空 tensor ✓
   - reduction 参数支持 string/int/enum 形式 ✓

8. **代码质量**：清晰的函数结构，`_normalize_reduction` 和 `_check_tensors` 辅助函数设计良好

---

**总体评价**：代码质量优秀，实现正确，测试覆盖完整。可以合并。

**Review by Dongxu-H — APPROVED** [review]

**Review by Caeruleann — APPROVED** [review]
